File: packages/schema_runtime_dart/lib/src/schema/schema_parser.dart
import 'schema_models.dart';

class SchemaParseError {
  const SchemaParseError({required this.path, required this.message});

  final String path;
  final String message;

  @override
  String toString() => '$path: $message';
}

class SchemaParseResult<T> {
  const SchemaParseResult({required this.value, required this.errors});

  final T? value;
  final List<SchemaParseError> errors;

  bool get isValid => value != null && errors.isEmpty;
}

SchemaParseResult<ComponentNode> parseComponentNode(Map<String, Object?> json) {
  final errors = <SchemaParseError>[];
  final node = _parseComponentNode(json, path: r'$', errors: errors);
  if (node == null) {
    return SchemaParseResult(value: null, errors: errors);
  }
  return SchemaParseResult(value: node, errors: errors);
}

SchemaParseResult<FragmentSchema> parseFragmentSchema(
  Map<String, Object?> json,
) {
  final errors = <SchemaParseError>[];

  String? requiredString(String key) {
    final value = json[key];
    if (value is String && value.isNotEmpty) {
      return value;
    }
    errors.add(
      SchemaParseError(path: key, message: 'Expected non-empty string'),
    );
    return null;
  }

  final schemaVersion = requiredString('schemaVersion');
  final id = requiredString('id');
  final documentType = requiredString('documentType');

  final nodeRaw = json['node'];
  final node = nodeRaw is Map
      ? _parseComponentNode(
          Map<String, Object?>.from(nodeRaw.cast<String, Object?>()),
          path: 'node',
          errors: errors,
        )
      : null;
  if (node == null) {
    errors.add(
      const SchemaParseError(path: 'node', message: 'Expected object'),
    );
  }

  if (schemaVersion == null ||
      id == null ||
      documentType == null ||
      documentType != 'fragment' ||
      node == null) {
    if (documentType != null && documentType != 'fragment') {
      errors.add(
        const SchemaParseError(
          path: 'documentType',
          message: 'Expected "fragment"',
        ),
      );
    }
    return SchemaParseResult(value: null, errors: errors);
  }

  return SchemaParseResult(
    value: FragmentSchema(
      schemaVersion: schemaVersion,
      id: id,
      documentType: documentType,
      node: node,
    ),
    errors: errors,
  );
}

SchemaParseResult<ScreenSchema> parseScreenSchema(Map<String, Object?> json) {
  final errors = <SchemaParseError>[];

  String? requiredString(String key) {
    final value = json[key];
    if (value is String && value.isNotEmpty) {
      return value;
    }
    errors.add(
      SchemaParseError(path: key, message: 'Expected non-empty string'),
    );
    return null;
  }

  final schemaVersion = requiredString('schemaVersion');
  final id = requiredString('id');
  final documentType = requiredString('documentType');
  final product = requiredString('product');
  final themeId = requiredString('themeId');

  final service = json['service'] as String?;
  final themeMode = json['themeMode'] as String?;

  final rootRaw = json['root'];
  final root = rootRaw is Map
      ? _parseComponentNode(
          Map<String, Object?>.from(rootRaw.cast<String, Object?>()),
          path: 'root',
          errors: errors,
        )
      : null;
  if (root == null) {
    errors.add(
      const SchemaParseError(path: 'root', message: 'Expected object'),
    );
  }

  final actions = <String, ActionDefinition>{};
  final actionsRaw = json['actions'];
  if (actionsRaw is Map) {
    for (final entry in actionsRaw.entries) {
      if (entry.key is! String) continue;
      final value = entry.value;
      if (value is! Map) {
        errors.add(
          SchemaParseError(
            path: 'actions.${entry.key}',
            message: 'Expected object',
          ),
        );
        continue;
      }
      final actionType = value['type'];
      if (actionType is! String || actionType.isEmpty) {
        errors.add(
          SchemaParseError(
            path: 'actions.${entry.key}.type',
            message: 'Expected non-empty string',
          ),
        );
        continue;
      }

      Map<String, Object?>? eventProperties;
      final eventPropsRaw = value['eventProperties'];
      if (eventPropsRaw is Map) {
        eventProperties = Map<String, Object?>.from(
          eventPropsRaw.cast<String, Object?>(),
        );
      }

      actions[entry.key as String] = ActionDefinition(
        type: actionType,
        route: value['route'] as String?,
        formId: value['formId'] as String?,
        eventName: value['eventName'] as String?,
        eventProperties: eventProperties,
      );
    }
  }

  if (schemaVersion == null ||
      id == null ||
      documentType == null ||
      product == null ||
      themeId == null ||
      root == null) {
    return SchemaParseResult(value: null, errors: errors);
  }

  return SchemaParseResult(
    value: ScreenSchema(
      schemaVersion: schemaVersion,
      id: id,
      documentType: documentType,
      product: product,
      service: service,
      themeId: themeId,
      themeMode: themeMode,
      root: root,
      actions: actions,
    ),
    errors: errors,
  );
}

ComponentNode? _parseComponentNode(
  Map<String, Object?> json, {
  required String path,
  required List<SchemaParseError> errors,
}) {
  final type = json['type'];
  if (type is! String || type.isEmpty) {
    errors.add(
      SchemaParseError(
        path: '$path.type',
        message: 'Expected non-empty string',
      ),
    );
    return null;
  }

  final propsRaw = json['props'];
  final props = propsRaw is Map
      ? Map<String, Object?>.from(propsRaw.cast<String, Object?>())
      : const <String, Object?>{};

  final actionsRaw = json['actions'];
  final actions = <String, String>{};
  if (actionsRaw is Map) {
    for (final entry in actionsRaw.entries) {
      if (entry.key is String && entry.value is String) {
        actions[entry.key as String] = entry.value as String;
      }
    }
  }

  final slots = <String, List<SchemaNode>>{};
  final slotsRaw = json['slots'];
  if (slotsRaw is Map) {
    for (final entry in slotsRaw.entries) {
      if (entry.key is! String) continue;
      final childrenRaw = entry.value;
      if (childrenRaw is! List) {
        errors.add(
          SchemaParseError(
            path: '$path.slots.${entry.key}',
            message: 'Expected array',
          ),
        );
        continue;
      }

      final children = <SchemaNode>[];
      for (var index = 0; index < childrenRaw.length; index++) {
        final child = childrenRaw[index];
        if (child is! Map) {
          errors.add(
            SchemaParseError(
              path: '$path.slots.${entry.key}[$index]',
              message: 'Expected object',
            ),
          );
          continue;
        }
        final childMap = Map<String, Object?>.from(
          child.cast<String, Object?>(),
        );
        if (childMap.containsKey('ref') && !childMap.containsKey('type')) {
          final ref = childMap['ref'];
          if (ref is String && ref.isNotEmpty) {
            children.add(RefNode(ref: ref));
          } else {
            errors.add(
              SchemaParseError(
                path: '$path.slots.${entry.key}[$index].ref',
                message: 'Expected non-empty string',
              ),
            );
          }
          continue;
        }

        final parsed = _parseComponentNode(
          childMap,
          path: '$path.slots.${entry.key}[$index]',
          errors: errors,
        );
        if (parsed != null) {
          children.add(parsed);
        }
      }
      slots[entry.key as String] = children;
    }
  }

  final bind = json['bind'] as String?;
  final visibleWhenRaw = json['visibleWhen'];
  final visibleWhen = visibleWhenRaw is Map
      ? Map<String, Object?>.from(visibleWhenRaw.cast<String, Object?>())
      : null;

  return ComponentNode(
    type: type,
    props: props,
    slots: slots,
    actions: actions,
    bind: bind,
    visibleWhen: visibleWhen,
  );
}

File: packages/schema_runtime_ts/src/schema/schema_parser.ts
import {
  type ActionDefinition,
  type ComponentNode,
  type FragmentSchema,
  type SchemaNode,
  type ScreenSchema,
  makeComponentNode,
  makeRefNode,
} from './schema_models.js';

export interface SchemaParseError {
  readonly path: string;
  readonly message: string;
}

export interface SchemaParseResult<T> {
  readonly value: T | null;
  readonly errors: SchemaParseError[];
  readonly isValid: boolean;
}

function result<T>(value: T | null, errors: SchemaParseError[]): SchemaParseResult<T> {
  return { value, errors, isValid: value !== null && errors.length === 0 };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function requiredNonEmptyString(
  json: Record<string, unknown>,
  key: string,
  errors: SchemaParseError[],
): string | null {
  const value = json[key];
  if (typeof value === 'string' && value.length > 0) return value;
  errors.push({ path: key, message: 'Expected non-empty string' });
  return null;
}

export function parseComponentNode(json: Record<string, unknown>): SchemaParseResult<ComponentNode> {
  const errors: SchemaParseError[] = [];
  const node = parseComponentNodeInternal(json, '$', errors);
  return node ? result(node, errors) : result<ComponentNode>(null, errors);
}

export function parseFragmentSchema(json: Record<string, unknown>): SchemaParseResult<FragmentSchema> {
  const errors: SchemaParseError[] = [];

  const schemaVersion = requiredNonEmptyString(json, 'schemaVersion', errors);
  const id = requiredNonEmptyString(json, 'id', errors);
  const documentType = requiredNonEmptyString(json, 'documentType', errors);

  const nodeRaw = json['node'];
  const node = isRecord(nodeRaw) ? parseComponentNodeInternal(nodeRaw, 'node', errors) : null;
  if (!node) {
    errors.push({ path: 'node', message: 'Expected object' });
  }

  if (!schemaVersion || !id || !documentType || documentType !== 'fragment' || !node) {
    if (documentType && documentType !== 'fragment') {
      errors.push({ path: 'documentType', message: 'Expected "fragment"' });
    }
    return result<FragmentSchema>(null, errors);
  }

  return result(
    {
      schemaVersion,
      id,
      documentType,
      node,
    },
    errors,
  );
}

export function parseScreenSchema(json: Record<string, unknown>): SchemaParseResult<ScreenSchema> {
  const errors: SchemaParseError[] = [];

  const schemaVersion = requiredNonEmptyString(json, 'schemaVersion', errors);
  const id = requiredNonEmptyString(json, 'id', errors);
  const documentType = requiredNonEmptyString(json, 'documentType', errors);
  const product = requiredNonEmptyString(json, 'product', errors);
  const themeId = requiredNonEmptyString(json, 'themeId', errors);

  const service = typeof json['service'] === 'string' ? (json['service'] as string) : undefined;
  const themeMode = typeof json['themeMode'] === 'string' ? (json['themeMode'] as string) : undefined;

  const rootRaw = json['root'];
  const root = isRecord(rootRaw) ? parseComponentNodeInternal(rootRaw, 'root', errors) : null;
  if (!root) {
    errors.push({ path: 'root', message: 'Expected object' });
  }

  const actions: Record<string, ActionDefinition> = {};
  const actionsRaw = json['actions'];
  if (isRecord(actionsRaw)) {
    for (const [key, rawValue] of Object.entries(actionsRaw)) {
      if (!isRecord(rawValue)) {
        errors.push({ path: `actions.${key}`, message: 'Expected object' });
        continue;
      }

      const actionType = rawValue['type'];
      if (typeof actionType !== 'string' || actionType.length === 0) {
        errors.push({ path: `actions.${key}.type`, message: 'Expected non-empty string' });
        continue;
      }

      const eventPropertiesRaw = rawValue['eventProperties'];
      const eventProperties = isRecord(eventPropertiesRaw)
        ? (eventPropertiesRaw as Record<string, unknown>)
        : undefined;

      actions[key] = {
        type: actionType,
        route: typeof rawValue['route'] === 'string' ? (rawValue['route'] as string) : undefined,
        formId: typeof rawValue['formId'] === 'string' ? (rawValue['formId'] as string) : undefined,
        modalId: typeof rawValue['modalId'] === 'string' ? (rawValue['modalId'] as string) : undefined,
        dataSource:
          typeof rawValue['dataSource'] === 'string' ? (rawValue['dataSource'] as string) : undefined,
        value: rawValue['value'],
        eventName:
          typeof rawValue['eventName'] === 'string' ? (rawValue['eventName'] as string) : undefined,
        eventProperties,
      };
    }
  }

  if (!schemaVersion || !id || !documentType || !product || !themeId || !root) {
    return result<ScreenSchema>(null, errors);
  }

  return result(
    {
      schemaVersion,
      id,
      documentType,
      product,
      service,
      themeId,
      themeMode,
      root,
      actions,
    },
    errors,
  );
}

function parseComponentNodeInternal(
  json: Record<string, unknown>,
  path: string,
  errors: SchemaParseError[],
): ComponentNode | null {
  const type = json['type'];
  if (typeof type !== 'string' || type.length === 0) {
    errors.push({ path: `${path}.type`, message: 'Expected non-empty string' });
    return null;
  }

  const propsRaw = json['props'];
  const props = isRecord(propsRaw) ? (propsRaw as Record<string, unknown>) : {};

  const actionsRaw = json['actions'];
  const actions: Record<string, string> = {};
  if (isRecord(actionsRaw)) {
    for (const [key, value] of Object.entries(actionsRaw)) {
      if (typeof value === 'string') actions[key] = value;
    }
  }

  const slots: Record<string, SchemaNode[]> = {};
  const slotsRaw = json['slots'];
  if (isRecord(slotsRaw)) {
    for (const [slotName, childrenRaw] of Object.entries(slotsRaw)) {
      if (!Array.isArray(childrenRaw)) {
        errors.push({ path: `${path}.slots.${slotName}`, message: 'Expected array' });
        continue;
      }

      const children: SchemaNode[] = [];
      for (let index = 0; index < childrenRaw.length; index++) {
        const child = childrenRaw[index];
        if (!isRecord(child)) {
          errors.push({ path: `${path}.slots.${slotName}[${index}]`, message: 'Expected object' });
          continue;
        }

        // Ref node (must NOT have type).
        if ('ref' in child && !('type' in child)) {
          const refValue = child['ref'];
          if (typeof refValue === 'string' && refValue.length > 0) {
            children.push(makeRefNode(refValue));
          } else {
            errors.push({
              path: `${path}.slots.${slotName}[${index}].ref`,
              message: 'Expected non-empty string',
            });
          }
          continue;
        }

        const parsed = parseComponentNodeInternal(child, `${path}.slots.${slotName}[${index}]`, errors);
        if (parsed) children.push(parsed);
      }

      slots[slotName] = children;
    }
  }

  const bind = typeof json['bind'] === 'string' ? (json['bind'] as string) : undefined;

  const visibleWhenRaw = json['visibleWhen'];
  const visibleWhen = isRecord(visibleWhenRaw)
    ? (visibleWhenRaw as Record<string, unknown>)
    : undefined;

  return makeComponentNode({
    type,
    props,
    slots,
    actions,
    bind,
    visibleWhen,
  });
}
